samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,1120 @@
+
+	VkClearValue black;
+	black.color = {{0, 0, 0, 1.0f}};
+
+	VkClearValue gray;
+	gray.color = {{0.5f, 0.5f, 0.5f, 1.0f}};
+
+	VkClearValue colorful;
+	colorful.color = {{
+	    (frame_number % 256) / 255.0f,
+	    ((frame_number + 63) % 256) / 255.0f,
+	    ((frame_number + 128) % 256) / 255.0f,
+	    1.0f,
+	}};
+
+	VkRenderPassBeginInfo rp_begin{VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO};
+	rp_begin.renderPass        = render_pass;
+	rp_begin.framebuffer       = swapchain_objects.framebuffers[index];
+	rp_begin.renderArea.extent = swapchain_extents;
+	rp_begin.clearValueCount   = 1;
+	rp_begin.pClearValues      = &black;
+	vkCmdBeginRenderPass(cmd, &rp_begin, VK_SUBPASS_CONTENTS_INLINE);
+
+	VkClearAttachment gray_clear;
+	gray_clear.aspectMask      = VK_IMAGE_ASPECT_COLOR_BIT;
+	gray_clear.colorAttachment = 0;
+	gray_clear.clearValue      = gray;
+
+	VkClearAttachment colorful_clear = gray_clear;
+	colorful_clear.clearValue        = colorful;
+
+	uint32_t half_width  = swapchain_extents.width / 2;
+	uint32_t half_height = swapchain_extents.height / 2;
+
+	VkClearRect gray_rect;
+	gray_rect.rect           = {{static_cast<int32_t>(half_width) / 2, static_cast<int32_t>(half_height)},
+                      {half_width, half_height / 2}};
+	gray_rect.baseArrayLayer = 0;
+	gray_rect.layerCount     = 1;
+
+	VkClearRect colorful_rect = gray_rect;
+	colorful_rect.rect        = {{250, 150}, {300, 300}};
+
+	vkCmdClearAttachments(cmd, 1, &gray_clear, 1, &gray_rect);
+	vkCmdClearAttachments(cmd, 1, &colorful_clear, 1, &colorful_rect);
+
+	vkCmdEndRenderPass(cmd);
+	VK_CHECK(vkEndCommandBuffer(cmd));
+
+	// Make a submission. Wait on the acquire semaphore and signal the present semaphore.
+	VkPipelineStageFlags wait_stage{VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT};
+
+	VkSubmitInfo info{VK_STRUCTURE_TYPE_SUBMIT_INFO};
+	info.commandBufferCount   = 1;
+	info.pCommandBuffers      = &cmd;
+	info.waitSemaphoreCount   = 1;
+	info.pWaitSemaphores      = &frame.acquire_semaphore;
+	info.pWaitDstStageMask    = &wait_stage;
+	info.signalSemaphoreCount = 1;
+	info.pSignalSemaphores    = &frame.present_semaphore;
+	VK_CHECK(vkQueueSubmit(queue->get_handle(), 1, &info, frame.submit_fence));
+}
+
+/**
+ * @brief Acquires an image from the swapchain.
+ * @param[out] index The swapchain index for the acquired image.
+ * @returns Vulkan result code
+ */
+VkResult SwapchainRecreation::acquire_next_image(uint32_t *index)
+{
+	PerFrame &frame = submit_history[submit_history_index];
+
+	// Use a fence to know when acquire is done.  Without VK_EXT_swapchain_maintenance1, this
+	// fence is used to infer when the _previous_ present to this image index has finished.
+	// There is no need for this with VK_EXT_swapchain_maintenance1.
+	VkFence acquire_fence = has_maintenance1 ? VK_NULL_HANDLE : get_fence();
+
+	VkResult result = vkAcquireNextImageKHR(get_device_handle(), swapchain, UINT64_MAX, frame.acquire_semaphore, acquire_fence, index);
```

Comment:
How does this prevent frame.acquire_semaphore from being 'in-use' at the time of the call to `vkAcquireNextImageKHR`?

Reply:
I'm not sure I understood the question. Are you concerned that acquire_semaphore may be used elsewhere at the same time this call is being made? That can't happen due to CPU throttling. See `setup_frame`. The semaphores are being recycled (instead of newed/deleted), but they are always recycled when they could have otherwise been deleted.

Reply:
> If semaphore is not VK_NULL_HANDLE it must not have any uncompleted signal or wait operations pending
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkAcquireNextImageKHR.html

Ahhh thats where the fence is. I was wondering where the 'wait for previous frame to complete is' and its there. 
Thanks for clarifying!